api/websocket.go

```diff
@@ -112,10 +114,18 @@ func (wsSvr *WebsocketHandler) handleConnection(ctx context.Context, ws *websock
 		return nil
 	})
 
-	ctx, cancel := context.WithCancel(ctx)
+	ctx, ctxCancel := context.WithCancel(WithStreamContext(ctx))
 	safeWs := &safeWebsocketConn{ws: ws}
-	go ping(ctx, safeWs, cancel)
-
+	go ping(ctx, safeWs, ctxCancel)
+
+	cancel := func() {
+		ctxCancel()
+		// clean up the stream context
+		sc, _ := StreamFromContext(ctx)
+		for _, id := range sc.ListenerIDs() {
+			wsSvr.coreService.ChainListener().RemoveResponder(id)
+		}
+	}
 	for {
 		select {
 		case <-ctx.Done():
```

Comment:
here should also call `cancel` to clean up (remove all responder)?

Reply:
good catch